extensions/env-injector/src/main.rs

```diff
@@ -0,0 +1,161 @@
+use http_body_util::{BodyExt, Full};
+use hyper::body::Bytes;
+use hyper::{Method, Request};
+use hyper_util::client::legacy::Client;
+use hyperlocal::{UnixClientExt, UnixConnector, Uri};
+use std::error::Error;
+use std::process::Command;
+use std::collections::HashMap;
+use std::path::Path;
+use std::io::BufRead;
+
+const SNAPD_SOCKET: &str = "/run/snapd-snap.socket";
+
+async fn snapdapi_req() -> Result<serde_json::Value, Box<dyn Error + Send + Sync>> {
+    let url: hyperlocal::Uri = Uri::new(SNAPD_SOCKET, "/v2/snapctl").into();
+
+    let client: Client<UnixConnector, Full<Bytes>> = Client::unix();
+
+    let snap_context = std::env::var("SNAP_CONTEXT")?;
+
+    let request_body = format!(
+        r#"{{"context-id":"{}","args":["get", "env", "envfile", "apps"]}}"#,
+        snap_context
+    );
+
+    let req: Request<Full<Bytes>> = Request::builder()
+        .method(Method::POST)
+        .uri(url)
+        .body(Full::from(request_body))?;
+
+    let mut res = client.request(req).await?;
+
+    let mut body: Vec<u8> = Vec::new();
+
+    while let Some(frame_result) = res.frame().await {
+        let frame = frame_result?;
+
+        if let Some(segment) = frame.data_ref() {
+            body.extend_from_slice(segment);
+        }
+    }
+
+    Ok(serde_json::from_slice(&body)?)
+}
+
+fn set_env_vars(app: &str, json: &serde_json::Value) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
+    let stdout_str = json["result"]["stdout"].as_str().ok_or("Invalid stdout")?;
+    let stdout_json: serde_json::Value = serde_json::from_str(stdout_str)?;
+
+    fn process_env(env: &serde_json::Value) -> HashMap<String, String> {
+        env.as_object()
+            .unwrap()
+            .iter()
+            .map(|(k, v)| {
+                let key = k.to_uppercase().replace("-", "_");
+                (key, v.to_string())
+            })
+            .collect()
+    }
+
+    if let Some(global_env) = stdout_json["env"].as_object() {
+        for (key, value) in process_env(&serde_json::Value::Object(global_env.clone())) {
+            std::env::set_var(key, value.trim_matches('"'));
+        }
+    }
+
+    if let Some(app_env) = stdout_json["apps"][app]["env"].as_object() {
+        for (key, value) in process_env(&serde_json::Value::Object(app_env.clone())) {
+            std::env::set_var(key, value.trim_matches('"'));
+        }
+    }
```

Comment:
This also processes env keys that contain dots, treated as a hierarchy by snapd.
Setting a snap option such as `env.x.bad=BAD` results in an unwanted export for `X={"bad":"test"}`.

Reply:
Fixed! Thank you!